Additions to author and reviewer guidance, and language code review checklists (#582)
microsoft/code-with-engineering-playbook: code-reviews/recipes/CSharp.md

* [ ] Does this code properly validate arguments sanity (i.e. [CA1062](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1062))? Consider leveraging extensions such as [Ensure.That](https://github.com/danielwertheim/Ensure.That)
* [ ] Does this code include telemetry ([metrics, tracing](https://docs.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview) and [logging](https://serilog.net/)) instrumentation?
* [ ] Does this code leverage the [options design pattern](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-3.1) by using classes to provide strongly typed access to groups of related settings?

> * [ ] Are constants used in the main class? Or if this is used across files/classes, is there a static class for the constants?

Reviewer: What do you mean by the first question (constants in main class)?

Author: ah, instead of using just raw strings (e.g. for routes or paths) in code :). Let me see if there is a better way to phrase this.

Author: Should be more clear now, thank you!

> * [ ] Are magic numbers explained? There should be no number in the code without at least a comment of why this is here. If the number is repetitive, is there a constant/enum or equivalent?

Reviewer: ```suggestion
* [ ] Are magic numbers explained? There should be no number in the code without at least a comment of why it is there. If the number is repetitive, is there a constant/enum or equivalent?
```

> * [ ] Is proper Exception handling set up? Catch(Exception) is not the right pattern, only exceptions that could happen need to be caught, e.g. `Catch(Exception e) when (e is IOException || e is ArgumentException)`.

Reviewer: ```suggestion
* [ ] Is proper exception handling set up? Catching the exception base class (`catch (Exception)`) is generally not the right pattern. Instead catch the specific exceptions that can happen e.g., `IOException`.
```

* [ ] Is the use of #pragma fair?
* [ ] Are tests arranged correctly with the Arrange/Act/Assert pattern and properly documented in this way?

> * [ ] If there is an async function, does the name of the function have Async in it?

Reviewer: ```suggestion
* [ ] If there is an asynchronous method, does the name of the method end with the `Async` suffix?
```

Reviewer: Some of these points go into small code conventions related details. While I have no strong preference, I can't but wonder if these could be combined into a single bullet point ("does the code comply with coding conventions") with a link to the conventions.

> * [ ] If a method is async, is Task.Delay used instead of Thread.Sleep? Task.Delay is not blocking the current thread and creates a task that will complete without blocking the thread, so in a multi thread, multi task environment, this is the one to prefer.

Reviewer: ```suggestion
* [ ] If a method is asynchronous, is `Task.Delay` used instead of `Thread.Sleep`? `Task.Delay` is not blocking the current thread and creates a task that will complete without blocking the thread, so in a multi-threaded, multi-task environment, this is the one to prefer.
```

* [ ] Is a cancellation token for async tasks needed rather than bool patterns?

> * [ ] Is a minimum level of logging in place? Is the logging level is the right one?

Reviewer: ```suggestion
* [ ] Is a minimum level of logging in place? Are the logging levels used sensible?
```

* [ ] Are internal vs private vs public classes and methods used the right way?
* [ ] Are auto property set and get used the right way? In a model without constructor and for deserialization, it is ok to have all accessible. But for other classes usually a private set or internal set is better.

> * [ ] Is the `using` pattern for streams and other disposable classes used? If not, better to have the Dispose method called explicitly.

Reviewer: ```suggestion
* [ ] Is the `using` pattern for streams and other disposable classes used? If not, better to have the `Dispose` method called explicitly.
```
